src/baxter_actions.cpp

```diff
@@ -0,0 +1,155 @@
+#include "rapid_pbd/baxter_actions.h"
+
+#include <string>
+#include <typeinfo>
+
+#include "actionlib/client/simple_action_client.h"
+#include "actionlib/server/simple_action_server.h"
+#include "control_msgs/FollowJointTrajectoryAction.h"
+#include "control_msgs/GripperCommandAction.h"
+#include "control_msgs/SingleJointPositionAction.h"
+#include "controller_manager_msgs/ListControllers.h"
+#include "controller_manager_msgs/SwitchController.h"
+
+#include "rapid_pbd_msgs/Action.h"
+#include "rapid_pbd_msgs/ArmControllerState.h"
+#include "rapid_pbd_msgs/FreezeArm.h"
+#include "rapid_pbd_msgs/RelaxArm.h"
+
+using actionlib::SimpleClientGoalState;
+using control_msgs::FollowJointTrajectoryFeedback;
+using control_msgs::FollowJointTrajectoryResult;
+using control_msgs::GripperCommandFeedback;
+using control_msgs::GripperCommandResult;
+using control_msgs::SingleJointPositionFeedback;
+using control_msgs::SingleJointPositionGoal;
+using control_msgs::SingleJointPositionResult;
+namespace msgs = rapid_pbd_msgs;
+
+namespace rapid {
+namespace pbd {
+namespace baxter {
+GripperAction::GripperAction(const std::string& action_name,
+                             const std::string& baxter_action_name)
+    : server_(action_name, boost::bind(&GripperAction::Execute, this, _1),
+              false),
+      baxter_client_(baxter_action_name, true) {}
+
+void GripperAction::Start() {
+  while (!baxter_client_.waitForServer(ros::Duration(5))) {
+    ROS_WARN("Waiting for Baxter gripper server to come up.");
+  }
+  server_.start();
+}
+
+void GripperAction::Execute(
+    const control_msgs::GripperCommandGoalConstPtr& goal) {
+  control_msgs::GripperCommandGoal baxter_goal;
+  // Baxter uses 0-100% corresponding to 0-0.1m
+  baxter_goal.command.position = goal->command.position * 1000;
+  baxter_goal.command.max_effort = goal->command.max_effort;
+  baxter_client_.sendGoal(
+      baxter_goal,
+      boost::function<void(const SimpleClientGoalState&,
+                           const GripperCommandResult::ConstPtr&)>(),
+      boost::function<void()>(),
+      boost::bind(&GripperAction::HandleFeedback, this, _1));
+  while (!baxter_client_.getState().isDone()) {
+    if (server_.isPreemptRequested() || !ros::ok()) {
+      baxter_client_.cancelAllGoals();
+      server_.setPreempted();
+      return;
+    }
+    ros::spinOnce();
+  }
+  if (baxter_client_.getState() == SimpleClientGoalState::PREEMPTED) {
+    baxter_client_.cancelAllGoals();
+    server_.setPreempted();
+    return;
+  } else if (baxter_client_.getState() == SimpleClientGoalState::ABORTED) {
+    baxter_client_.cancelAllGoals();
+    server_.setAborted();
+    return;
+  }
+
+  GripperCommandResult::ConstPtr baxter_result = baxter_client_.getResult();
+  control_msgs::GripperCommandResult result;
```

Comment:
Is there anything different here? Can we just return `baxter_result`?

Reply:
I can't pass it directly as they are different types